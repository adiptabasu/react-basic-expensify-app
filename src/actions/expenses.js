import { v4 as uuid } from 'uuid';
import db, { push, ref, onValue, get } from '../firebase/firebase';
const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = {
            description,
            note,
            amount,
            createdAt
        }
        return push(ref(db, 'expenses'), expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        });
    };
};

const removeExpense = ({ id = undefined } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

const editExpense = ({ id, updates }) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

const startSetExpenses = (expensesData = {}) => {
    return (dispatch) => {
        return onValue(ref(db, 'expenses'), (snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnap) => {
                expenses.push({
                    id: childSnap.key,
                    ...childSnap.val()
                })
            })
            dispatch(setExpenses(expenses));
            // const setStartExpensePromise = new Promise((resolve, reject) => {
            //     resolve(dispatch(setExpenses(expenses)));
            // });
            // console.log('promise', setStartExpensePromise);
            // return setStartExpensePromise;
        }, { onlyOnce: true }, (e) => {
            console.log('Error is ', e);
        });
    };
};

// const startSetExpenses = (dispatch) => {
//     return (dispatch) => {
//         const dbRef = ref(db)
//         return get(child(dbRef, 'expense')).then((snapshot) => {
//             if (snapshot.exists()) {
//                 const expenses = []
//                 snapshot.forEach((childSnapshot) => {
//                     expenses.push({
//                         id: childSnapshot.key,
//                         ...childSnapshot.val(),
//                     });
//                 });
//                 dispatch(setExpenses(expenses))
//             } else {
//                 console.log("No data available");
//             }
//         }).catch((error) => {
//             console.error(error);
//         });
//     }
// }

export { addExpense, removeExpense, editExpense, setExpenses, startSetExpenses };