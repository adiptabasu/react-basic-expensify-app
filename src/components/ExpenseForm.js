import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
// import '../../node_modules/react-dates/lib/css/_datepicker.css';
const now = moment();

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ''
        }
        this.onDescriptionChange = this.onDescriptionChange.bind(this);
        this.onNoteChange = this.onNoteChange.bind(this);
        this.onAmountChange = this.onAmountChange.bind(this);
        this.onDateChange = this.onDateChange.bind(this);
        this.onCalendarFocusChange = this.onCalendarFocusChange.bind(this);
        this.onFormSubmit = this.onFormSubmit.bind(this);
    }
    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        const matchVal = amount.match(/^\d+\.{0,1}\d{0,2}$/);
        if (!!matchVal || !amount)
            this.setState(() => ({ amount }));
        else
            this.setState((prevState) => {
                amount: prevState.amount
            })
    }
    onDateChange = (createdAt) => {
        if (createdAt)
            this.setState(() => ({ createdAt }));
    }
    onCalendarFocusChange = ({ focused }) => {
        this.setState({ calenderFocused: focused });
    }
    onFormSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Description and Amount are Mandatory' }));
        }
        else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onFormSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.onCalendarFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => (false)}
                    />
                    <textarea
                        placeholder="Add a Note"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;