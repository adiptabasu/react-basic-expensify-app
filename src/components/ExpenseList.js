import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectedExpenses from '../selectors/expenses';

const ExpenseList = (props) => (
    <div>
        {props.expenses.length == 0 ? (
            <p>No expenses</p>
        ) : (
                props.expenses.map((expense) => (
                    <ExpenseListItem
                        key={expense.id}
                        {...expense} />
                ))
            )}
    </div>
);

const mapStateToProps = (state) => {
    return {
        // expenses: state.expenses,
        // filters: state.filters
        expenses: selectedExpenses(state.expenses, state.filters)
    }
}

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

export default ConnectedExpenseList;
export { ExpenseList };