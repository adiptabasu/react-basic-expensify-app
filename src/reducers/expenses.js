const expencesReducerDefaultState = [];
const expencesReducer = (state = expencesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE': {
            // console.log(action.expense)
            // return state.concat(action.expense);
            return [...state, action.expense];
            break;
        }
        case 'REMOVE_EXPENSE': {
            return state.filter((expense) => expense.id != action.id);
            break;
        }
        case 'EDIT_EXPENSE': {
            return state.map((expense) => {
                if (expense.id == action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                }
                else {
                    return expense;
                }
            })
            break;
        }
        default: {
            return state;
            break;
        }
    }
};

export default expencesReducer;