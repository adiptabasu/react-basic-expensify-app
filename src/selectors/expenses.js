import moment from 'moment';

const getVisibleExpences = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        createdAtMoment.isSameOrBefore
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = text.trim().length <= 0 ? true : (expense.description.toLowerCase().includes(text.toLowerCase()));
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy == 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        else if (sortBy == 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};

export default getVisibleExpences;