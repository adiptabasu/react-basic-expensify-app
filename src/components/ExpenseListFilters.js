import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            calendarFocused: null
        }
        this.onDatesChange = this.onDatesChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate({ startDate }));
        this.props.dispatch(setEndDate({ endDate }));
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            className="text-input"
                            type="text"
                            value={this.props.filters.text} onChange={(e) => {
                                // console.log(e.target.value);
                                this.props.dispatch(setTextFilter({ text: e.target.value }));
                            }}
                            placeholder="Search Expenses"
                        />
                    </div>
                    <div className="input-group__item">
                        <select className="select" value={this.props.sortBy} onChange={(e) => {
                            let sortVal = e.target.value;
                            if (sortVal == 'date') {
                                this.props.dispatch(sortByDate());
                            }
                            else if (sortVal == 'amount') {
                                this.props.dispatch(sortByAmount());
                            }
                        }}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => (false)}
                            showClearDates={true}
                        />
                    </div>
                </div>


            </div>
        );
    }
}

// const ExpenseListFilters = (props) => (
//     <div>
//         <input type="text" value={props.filters.text} onChange={(e) => {
//             // console.log(e.target.value);
//             props.dispatch(setTextFilter({ text: e.target.value }));
//         }} />
//         <select value={props.sortBy} onChange={(e) => {
//             let sortVal = e.target.value;
//             if (sortVal == 'date') {
//                 props.dispatch(sortByDate());
//             }
//             else if (sortVal == 'amount') {
//                 props.dispatch(sortByAmount());
//             }
//         }}>
//             <option value="date">Date</option>
//             <option value="amount">Amount</option>
//         </select>
//     </div>
// );

const mapStateToProps = (state) => {
    // console.log(state.filters);
    return { filters: state.filters }
}

const ConnectedExpenseListFilters = connect(mapStateToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;