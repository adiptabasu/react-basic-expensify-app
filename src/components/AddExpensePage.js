import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import { useNavigate } from 'react-router-dom';

const AddExpensePage = (props) => {
    const navigate = useNavigate();
    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Add Expense</h1>
                </div>
            </div>
            <div className="content-container">
                <ExpenseForm
                    onSubmit={(expense) => {
                        // console.log('Inside AddExpensePage');
                        // props.dispatch(addExpense(expense));
                        props.onSubmit(expense);
                        navigate('/');
                    }}
                />
            </div>
        </div>
    );
};

const mapDispatchToProps = (dispatch) => (({ onSubmit: (expense => dispatch(startAddExpense(expense))) }))

const ConnectedAddExpensePage = connect(undefined, mapDispatchToProps)(AddExpensePage);

export default ConnectedAddExpensePage;