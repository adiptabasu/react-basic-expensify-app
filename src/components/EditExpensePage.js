import React from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRomoveExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
    let params = useParams();
    const navigate = useNavigate();
    const expenseSelected = props.expenses.find((exp) => (exp.id == params.id));
    console.log(expenseSelected);
    return (<div>
        <ExpenseForm
            expense={expenseSelected}
            onSubmit={(expense) => {
                console.log('updated', expense);
                props.dispatch(startEditExpense({ id: params.id, updates: expense }));
                navigate('/');
            }}
        />
        <button onClick={() => {
            props.dispatch(startRomoveExpense({ id: params.id }));
            navigate('/');
        }}>Remove</button>
    </div>);
};

const mapStateToProps = (state) => {
    return ({
        expenses: state.expenses
    });
}

const ConnectedEditExpensePage = connect(mapStateToProps)(EditExpensePage);

export default ConnectedEditExpensePage;