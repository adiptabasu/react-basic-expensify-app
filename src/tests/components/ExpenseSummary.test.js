import { test, expect } from '@jest/globals';
import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';

test('should correctly render Expense Summary with one expense', () => {
    const wrapper = shallow(<ExpenseSummary
        expenseCount={1}
        expensesTotal={235}
    />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render Expense Summary with multiple expenses', () => {
    const wrapper = shallow(<ExpenseSummary
        expenseCount={2}
        expensesTotal={735}
    />);
    expect(wrapper).toMatchSnapshot();
});