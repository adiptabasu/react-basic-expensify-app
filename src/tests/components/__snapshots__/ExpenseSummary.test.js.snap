// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render Expense Summary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totaling 
    $7.35
  </h1>
</div>
`;

exports[`should correctly render Expense Summary with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $2.35
  </h1>
</div>
`;
